ipa-core/src/protocol/basics/mul/dzkp_malicious.rs

```diff
@@ -80,20 +80,143 @@ impl<'a, F: Field + DZKPCompatibleField<N>, const N: usize>
+            F::as_segment_entry(b.right_arr()),
+            F::as_segment_entry(prss.left_arr()),
+            F::as_segment_entry(prss.right_arr()),
+            F::as_segment_entry(z.right_arr()),
+        );
+
+        // add segment to the batch that needs to be verified by the dzkp prover and verifiers
+        ctx.push(record_id, segment);
+
+        Ok(z)
+    }
+    fn generate_share_from_three_bits(role: Role, i: usize) -> AdditiveShare<Boolean> {
+        let (first_bit, second_bit) = match role {
+            Role::H1 => (i % 2 == 0, (i >> 1) % 2 == 0),
+            Role::H2 => ((i >> 1) % 2 == 0, (i >> 2) % 2 == 0),
+            Role::H3 => ((i >> 2) % 2 == 0, i % 2 == 0),
+        };
+        <AdditiveShare<Boolean>>::from((first_bit.into(), second_bit.into()))
+    }
+
+    fn all_combination_of_inputs(role: Role, i: usize) -> [AdditiveShare<Boolean>; 3] {
+        // first three bits
+        let a = generate_share_from_three_bits(role, i);
+        // middle three bits
+        let b = generate_share_from_three_bits(role, i >> 3);
+        // last three bits
+        let prss = generate_share_from_three_bits(role, i >> 6);
+
+        [a, b, prss]
+    }
+
+    #[derive(CompactStep)]
+    enum TestStep {
+        #[step(count = 512)]
+        Counter(usize),
+    }
+
+    #[tokio::test]
+    async fn detect_cheating() {
+        let world = TestWorld::default();
+
+        for i in 0..512 {
+            let [(_, s_1), (_, s_2), (v_3, s_3)] = world
+                .malicious((), |ctx, ()| async move {
+                    let [a, b, prss] = all_combination_of_inputs(ctx.role(), i);
+                    let validator = ctx.narrow(&TestStep::Counter(i)).dzkp_validator(10);
+                    let mctx = validator.context();
+                    let product = multiply_with_cheater(
+                        mctx.set_total_records(1),
+                        RecordId::FIRST,
+                        &a,
+                        &b,
+                        &prss,
+                        H1,
+                    )
+                    .await
+                    .unwrap();
+
+                    (
+                        validator.validate().await,
+                        [
+                            bool::from(*a.left_arr().first()),
+                            bool::from(*a.right_arr().first()),
+                            bool::from(*b.left_arr().first()),
+                            bool::from(*b.right_arr().first()),
+                            bool::from(*prss.left_arr().first()),
+                            bool::from(*prss.right_arr().first()),
+                            bool::from(*product.left_arr().first()),
+                            bool::from(*product.right_arr().first()),
+                        ],
+                    )
+                })
+                .await;
+
+            // H1 cheats means H3 fails
+            // since always verifier on the left of the cheating prover fails
```

Comment:
This makes zero sense to me. I cannot think of any reason why just one verifier should fail.

Both verifiers should be checking that all of the `b` shares sum to zero. If it fails for one, it should fail for both.

Reply:
lets sync on this once you are back next week.